File: sdks/node/tests/buffer.test.ts
import { RingBuffer } from '../src/buffer';

describe('RingBuffer', () => {
  const makeEvent = (id: number) => ({
    customerId: `cust_${id}`,
    metricName: 'api_calls',
    quantity: 1,
    idempotencyKey: `key_${id}`,
    occurredAt: new Date().toISOString(),
  });

  it('should push and drain events', () => {
    const buf = new RingBuffer(10);
    buf.push(makeEvent(1));
    buf.push(makeEvent(2));
    buf.push(makeEvent(3));

    expect(buf.size).toBe(3);
    expect(buf.isEmpty).toBe(false);

    const drained = buf.drain();
    expect(drained).toHaveLength(3);
    expect(drained[0].customerId).toBe('cust_1');
    expect(drained[2].customerId).toBe('cust_3');
    expect(buf.size).toBe(0);
    expect(buf.isEmpty).toBe(true);
  });

  it('should drop oldest on overflow', () => {
    const buf = new RingBuffer(3);
    expect(buf.push(makeEvent(1))).toBe(true);
    expect(buf.push(makeEvent(2))).toBe(true);
    expect(buf.push(makeEvent(3))).toBe(true);
    expect(buf.isFull).toBe(true);

    // Overflow — drops event 1
    expect(buf.push(makeEvent(4))).toBe(false);
    expect(buf.size).toBe(3);

    const drained = buf.drain();
    expect(drained[0].customerId).toBe('cust_2');
    expect(drained[1].customerId).toBe('cust_3');
    expect(drained[2].customerId).toBe('cust_4');
  });

  it('should drainUpTo a max number of events', () => {
    const buf = new RingBuffer(100);
    for (let i = 0; i < 10; i++) buf.push(makeEvent(i));

    const batch1 = buf.drainUpTo(3);
    expect(batch1).toHaveLength(3);
    expect(batch1[0].customerId).toBe('cust_0');
    expect(buf.size).toBe(7);

    const batch2 = buf.drainUpTo(5);
    expect(batch2).toHaveLength(5);
    expect(batch2[0].customerId).toBe('cust_3');
    expect(buf.size).toBe(2);
  });

  it('should handle drain on empty buffer', () => {
    const buf = new RingBuffer(10);
    expect(buf.drain()).toEqual([]);
    expect(buf.drainUpTo(5)).toEqual([]);
  });

  it('should throw on invalid capacity', () => {
    expect(() => new RingBuffer(0)).toThrow('capacity must be >= 1');
    expect(() => new RingBuffer(-5)).toThrow('capacity must be >= 1');
  });

  it('should handle wrap-around correctly', () => {
    const buf = new RingBuffer(4);
    buf.push(makeEvent(1));
    buf.push(makeEvent(2));
    buf.drainUpTo(2); // head wraps forward

    buf.push(makeEvent(3));
    buf.push(makeEvent(4));
    buf.push(makeEvent(5));
    buf.push(makeEvent(6)); // tail wraps around

    expect(buf.size).toBe(4);
    const drained = buf.drain();
    expect(drained.map(e => e.customerId)).toEqual([
      'cust_3', 'cust_4', 'cust_5', 'cust_6',
    ]);
  });
});

File: sdks/go/buffer_test.go
package metering

import "testing"

func makeEvent(n int) resolvedEvent {
	return resolvedEvent{
		CustomerID:     "cust_" + string(rune('0'+n)),
		MetricName:     "api_calls",
		Quantity:       1,
		IdempotencyKey: "key",
		OccurredAt:     "2026-03-21T00:00:00Z",
	}
}

func TestRingBuffer_PushAndDrain(t *testing.T) {
	buf := newRingBuffer(10)
	buf.push(makeEvent(1))
	buf.push(makeEvent(2))
	buf.push(makeEvent(3))

	if buf.size() != 3 {
		t.Fatalf("expected size 3, got %d", buf.size())
	}

	items := buf.drain()
	if len(items) != 3 {
		t.Fatalf("expected 3 items, got %d", len(items))
	}
	if buf.size() != 0 {
		t.Fatalf("expected empty after drain, got %d", buf.size())
	}
}

func TestRingBuffer_Overflow(t *testing.T) {
	buf := newRingBuffer(3)
	buf.push(makeEvent(1))
	buf.push(makeEvent(2))
	buf.push(makeEvent(3))

	ok := buf.push(makeEvent(4))
	if ok {
		t.Fatal("expected overflow (false), got true")
	}
	if buf.size() != 3 {
		t.Fatalf("expected size 3 after overflow, got %d", buf.size())
	}

	items := buf.drain()
	if items[0].CustomerID != "cust_2" {
		t.Fatalf("expected oldest dropped, got %s", items[0].CustomerID)
	}
}

func TestRingBuffer_DrainUpTo(t *testing.T) {
	buf := newRingBuffer(100)
	for i := 0; i < 10; i++ {
		buf.push(makeEvent(i))
	}

	batch := buf.drainUpTo(3)
	if len(batch) != 3 {
		t.Fatalf("expected 3, got %d", len(batch))
	}
	if buf.size() != 7 {
		t.Fatalf("expected 7 remaining, got %d", buf.size())
	}
}

func TestRingBuffer_DrainEmpty(t *testing.T) {
	buf := newRingBuffer(10)
	items := buf.drain()
	if items != nil {
		t.Fatal("expected nil from empty drain")
	}
}

func TestRingBuffer_PanicOnZeroCapacity(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected panic on zero capacity")
		}
	}()
	newRingBuffer(0)
}
